core/unix/os.c

```diff
@@ -3474,10 +3474,17 @@ os_thread_suspend(thread_record_t *tr)
      */
     mutex_unlock(&ostd->suspend_lock);
     while (ksynch_get_value(&ostd->suspended) == 0) {
-        /* For Linux, waits only if the suspended flag is not set as 1. Return value
-         * doesn't matter because the flag will be re-checked.
+        /* For Linux, waits only if the suspended flag is not set as 1. Other
+         * than a timeout value, the return value doesn't matter because the
+         * flag will be re-checked.
          */
-        ksynch_wait(&ostd->suspended, 0, 0);
+        if (ksynch_wait(&ostd->suspended, 0, 5000) == -ETIMEDOUT) {
+            mutex_lock(&ostd->suspend_lock);
+            ASSERT(ostd->suspend_count > 0);
```

Comment:
I don't think this assert is valid: a resume could have happened after ksynch_wait returns but before the lock is acquired.

Reply:
I believe the contract of os_thread_suspend/os_thread_resume is that you cannot call into os_thread_resume more times than os_thread_suspend has returned. Therefore, even if other os_thread_suspend/os_thread_resume calls went through and modified suspend_count between the ksynch_wait and mutex_lock, since this was the thread which incremented suspend_count and hasn't returned yet, it would be incorrect for any caller to call os_thread_resume and cause suspend_count to be decremented down to 0.

Please correct me if my logic or assumptions are incorrect.

Reply:
Yes, this sounds right.